### PR TITLE
[ci] add coverage reporting

### PR DESCRIPTION
**Pull request by Saviq**

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
Needs rebasing (or a merge) to pick up #2381?

(This is the cost side of https://github.com/MirServer/wlcs/pull/231)

**Comment by AlanGriffiths** [comment]
Here's the split I'm thinking of:

1. run by "make [p]test"
   - GPL-required
   - LGPL-required
   - mir_acceptance_tests
   - miral-test
   - miral-test-internal
   - mir_integration_tests
   - mir_umock_acceptance_tests
   - mir_umock_unit_tests
   - mir_unit_tests
   - mir_unit_tests_eglstream-kms
   - mir_unit_tests_gbm-kms
   - mir_unit_tests_x11
   - package-abis
   - test_wayland_generator
   - unit-tests-do-not-link-with-GL-or-EGL
   - wlcs
   - wlcs_*

2. Run in CI lab
   - mir_performance_test
   - mir-smoke-test-runner

3. Run by manual testing
   -

**Comment by Saviq** [comment]
> Here's the split I'm thinking of:
> 
> 1. run by "make [p]test"
> 2. Run in CI lab
> 3. Run by manual testing

That's already the case, no? (with the exception that mir-smoke-test-runner [currently runs on `make test`](https://github.com/MirServer/mir/blob/a89b44bffca8b2e4df4522ac67564699c224950f/tests/performance-tests/CMakeLists.txt#L34-L38))

With the above I meant to further group the coverage measured by the type of tests that ran. IMO it's interesting to know whether something was covered by unit tests vs. integration tests?

See https://docs.codecov.com/docs/flags for how it gets used.

File: .github/workflows/coverage.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+          gcovr \
+          lcov \
+          ninja-build
+
+        mk-build-deps --root-cmd sudo --install
+
+    - name: Configure
+      run: >
+        cmake
+        -DCMAKE_BUILD_TYPE=Coverage
```
Comment (AlanGriffiths) [inline .github/workflows/coverage.yml:57]: As discussed in #2395
```suggestion
        -DCMAKE_BUILD_TYPE=Coverage
        -DMIR_RUN_PERFORMANCE_TESTS=ON        
```

**Comment by Saviq** [comment]
Another refresh, needs discussion on how we want to progress with this.

**Comment by Saviq** [comment]
Discrepancy in LOC:

```plain
$ cloc .
    1859 text files.
    1734 unique files.                                          
     144 files ignored.

github.com/AlDanial/cloc v 1.96  T=1.20 s (1440.7 files/s, 215196.8 lines/s)
--------------------------------------------------------------------------------
Language                      files          blank        comment           code
--------------------------------------------------------------------------------
C++                             694          25304          14077         124159
C/C++ Header                    835          10179          19310          37841
XML                              24           1768             43          11927
CMake                           125            905            393           4665
C                                 8            221            223           2242
Python                            7            224            538            953
Bourne Shell                     13            177             59            917
XSLT                              1             51            107            732
YAML                             11            120             40            658
Markdown                          9            198              0            457
CSS                               2             26              0            161
Bourne Again Shell                2             26              7            117
make                              1             20             19             97
SVG                               1              0              1             26
TOML                              1              0              0             13
--------------------------------------------------------------------------------
SUM:                           1734          39219          34817         184965
--------------------------------------------------------------------------------
```

**Comment by Saviq** [comment]
> Discrepancy in LOC:

This makes more sense:

```plain
cloc src include 
    1165 text files.
    1138 unique files.                                          
      27 files ignored.

github.com/AlDanial/cloc v 1.96  T=0.57 s (1981.3 files/s, 270146.3 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
C++                            395          11680           8385          64939
C/C++ Header                   657           8266          16324          29724
XML                             21           1543             30           8264
CMake                           55            383            110           2054
C                                6            143            188           1858
Python                           4            156            499            619
-------------------------------------------------------------------------------
SUM:                          1138          22171          25536         107458
-------------------------------------------------------------------------------
```

**Comment by Saviq** [comment]
> This makes more sense:

But not really, since currently the codecov reports also on `tests/*`.

**Comment by Saviq** [comment]
> Discrepancy in LOC:

Codecov doesn't track bracket-only lines, so there's really no point in comparing those numbers.

**Comment by Saviq** [comment]
@AlanGriffiths ready to go.

**Review by AlanGriffiths — APPROVED** [review]
OK, "Go"...

bors merge

**Pull request merged**